crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -667,18 +670,31 @@ impl<'db> TypeInferenceBuilder<'db> {
     }
 
     fn infer_name_expression(&mut self, name: &ast::ExprName) -> Type<'db> {
-        let ast::ExprName {
-            range: _,
-            id: _,
-            ctx,
-        } = name;
+        let ast::ExprName { range: _, id, ctx } = name;
 
         match ctx {
             ExprContext::Load => {
-                let use_def = self.index.use_def_map(self.scope.file_scope_id(self.db));
+                let file_scope_id = self.scope.file_scope_id(self.db);
+                let use_def = self.index.use_def_map(file_scope_id);
                 let use_id = name.scoped_use_id(self.db, self.scope);
                 let definitions = use_def.use_definitions(use_id);
-                definitions_ty(self.db, definitions, use_def.use_may_be_unbound(use_id))
+                let may_be_unbound = use_def.use_may_be_unbound(use_id);
+                let symbols = self.index.symbol_table(file_scope_id);
+                // SAFETY: the symbol table always creates a symbol for every Name node.
+                let symbol = symbols.symbol_by_name(id).unwrap();
+
+                let unbound_ty = if may_be_unbound {
+                    if !symbol.is_defined() || !self.scope.is_function_like(self.db) {
+                        // implicit global
+                        Some(module_global_symbol_ty_by_name(self.db, self.file, id))
+                    } else {
+                        Some(Type::Unbound)
+                    }
+                } else {
+                    None
+                };
+
+                definitions_ty(self.db, definitions, unbound_ty)
```

Comment:
I think it would help readability if we move some declarations closer to the blocks that use them rather than defining all of them upfront. 


```rust
let file_scope_id = self.scope.file_scope_id(self.db);
let use_def = self.index.use_def_map(file_scope_id);
let use_id = name.scoped_use_id(self.db, self.scope);
let definitions = use_def.use_definitions(use_id);
let may_be_unbound = use_def.use_may_be_unbound(use_id);


let unbound_ty = if may_be_unbound {
		let symbols = self.index.symbol_table(file_scope_id);
    // SAFETY: the symbol table always creates a symbol for every Name node.
    let symbol = symbols.symbol_by_name(id).unwrap();

    if !symbol.is_defined() || !self.scope.is_function_like(self.db) {
        // implicit global
        Some(module_global_symbol_ty_by_name(self.db, self.file, id))
    } else {
        Some(Type::Unbound)
    }
} else {
    None
};